execution_engine/src/core/runtime/externals.rs

```diff
@@ -1011,6 +1016,32 @@ where
                 let ret = self.load_authorization_keys(len_ptr, result_size_ptr)?;
                 Ok(Some(RuntimeValue::I32(api_error::i32_from(ret))))
             }
+
+            FunctionIndex::NextAddress => {
+                let (out_ptr, out_size) = Args::parse(args)?;
+                self.charge_host_function_call(
+                    &host_function_costs.next_address,
+                    [out_ptr, out_size],
+                )?;
+
+                let next_address = self
+                    .context()
+                    .address_generator()
+                    .try_borrow_mut()
+                    .map_err(|_| Error::AddressGenerationFailed)?
```

Comment:
I'm skeptical about the new variant for this purpose and using `try_borrow_mut` - in other places, we assume we never keep a borrowed address generator before `borrowing_mut` it. If somehow, later or earlier in the code, an `address_generator` is still borrowed, we would easily catch it in the tests. All the uses of address generators use a temporary `borrow_mut()` instance to avoid this problem.

We already have `RuntimeContext::new_hash_address,` which does exactly this (it uses KEY_HASH_LENGTH, which coincidentally is the same as ADDRESS_LENGTH, maybe it's worth using https://github.com/casper-network/casper-node/blob/033e00309661297ae877934da15a78dde8254df5/execution_engine/src/core.rs#L11-L15 for this? Or have a separate `new_address` method etc.)

Reply:
Ok, cool stuff. To stay consistent we can indeed drop the solution with `try_borrow_mut()` and just use the available `new_hash_address()` helper.

I think that a separate `new_address()` would be a bit of exaggeration, since I doubt that the `KEY_HASH_LENGTH` and the actual address length will ever diverge. After all, the comment for `new_hash_address()` says: `Generates new deterministic hash for uses as an address.`

Changed in e68216d2aca0c940c25dae6d4e83415ea7912cee